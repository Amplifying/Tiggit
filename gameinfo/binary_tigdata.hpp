#ifndef __TIGDATA_BINARY_HPP_
#define __TIGDATA_BINARY_HPP_

#include "tiglist.hpp"

namespace TigData
{
  void fromBinary(TigData::TigList &out, const std::string &file);
  void toBinary(const TigData::TigList &list, const std::string &file);
}
#endif
