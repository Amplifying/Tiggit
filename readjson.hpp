#ifndef __READ_JSON_HPP
#define __READ_JSON_HPP

// FIXME / TODO: This is VERY UNCLEAN, and only works as long as we
// only have one cpp file. Once this changes, this has to be
// outsourced to a cpp as well.

#include <json/json.h>
#include <fstream>
#include <stdexcept>

/*
Json::Value parseJsonString(const std::string &string)
{
  using namespace Json;

  Value root;

  Reader reader;
  if(!reader.parse(string, root))
    throw std::runtime_error(reader.getFormatedErrorMessages());

  return root;
}
*/

Json::Value readJson(const std::string &file)
{
  using namespace Json;

  std::ifstream inf(file.c_str());
  if(!inf)
    throw std::runtime_error("Cannot read " + file);

  Value root;

  Reader reader;
  if(!reader.parse(inf, root))
    throw std::runtime_error(reader.getFormatedErrorMessages() + " reading " + file);

  return root;
}

#endif
