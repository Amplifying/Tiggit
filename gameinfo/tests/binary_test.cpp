#include "json_tigdata.hpp"
#include "binary_tigdata.hpp"
#include <readjson/readjson.hpp>
#include <iostream>

using namespace std;
using namespace TigData;

struct Fetch : FetchTig
{
  std::string path;

  Json::Value fetchTig(const std::string &idname,
                       const std::string &url)
  {
    string file = path + idname + ".tig";
    return ReadJson::readJson(file);
  }
};

void load(TigList &list,
          const std::string &tigfile,
          const std::string &path)
{
  Fetch fetch;
  fetch.path = path;
  if(fetch.path[path.size()-1] != '/')
    fetch.path += '/';
  fromJson(list, ReadJson::readJson(tigfile), fetch);
}

void print(TigList &list)
{
  cout << "\nChannel: " << list.channel
       << "\nDescription: " << list.desc
       << "\nHomepage: " << list.homepage << endl;

  for(int i=0; i<list.list.size(); i++)
    {
      const TigEntry &ent = list.list[i];
      cout << "\n" << ent.urlname << ":\n";
      cout << "  Title: " << ent.tigInfo.title
           << "\n  Homepage: " << ent.tigInfo.homepage
           << "\n  Tags: " << ent.tigInfo.tags
           << "\n  Channel: " << ent.channel
           << "\n  Rating: " << ent.rating << " (" << ent.rateCount << " votes)\n";
      cout << "  Demo: " << (ent.tigInfo.isDemo?"yes":"no") << endl;
    }
}

int main(int argc, char**argv)
{
  std::string binfile = "_test.bin";
  if(argc < 2 || argc > 3)
    {
      cout << "Syntax: app tiglist.json tigdir/\n";
      cout << "- or -: app binfile.dat\n";
      return 1;
    }

  if(argc == 3)
    {
      TigList list;
      load(list, argv[1], argv[2]);
      toBinary(list, binfile);
    }
  else
    binfile = argv[1];

  TigList list;
  fromBinary(list, binfile);
  print(list);

  return 0;
}
