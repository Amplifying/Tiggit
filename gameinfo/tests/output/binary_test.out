Syntax: app tiglist.json tigdir/
- or -: app binfile.dat
